services/github/github-deployments.service.js

```diff
@@ -0,0 +1,122 @@
+'use strict'
+
+const gql = require('graphql-tag')
+const Joi = require('@hapi/joi')
+const { renderBuildStatusBadge } = require('../build-status')
+const { GithubAuthV4Service } = require('./github-auth-service')
+const { documentation, transformErrors } = require('./github-helpers')
+
+const schema = Joi.object({
+  data: Joi.object({
+    repository: Joi.object({
+      deployments: Joi.object({
+        nodes: Joi.array()
+          .items(
+            Joi.object({
+              latestStatus: Joi.object({
+                state: Joi.string(),
+              }),
+            })
+          )
+          .required(),
+      }).required(),
+    }).required(),
+  }).required(),
+}).required()
+
+module.exports = class GithubDeployments extends GithubAuthV4Service {
+  static get category() {
+    return 'other'
+  }
+
+  static get route() {
+    return {
+      base: 'github/deployments',
+      pattern: ':user/:repo/:environment',
+    }
+  }
+
+  static get examples() {
+    return [
+      {
+        title: 'GitHub deployments',
+        namedParams: {
+          user: 'badges',
+          repo: 'shields',
+          environment: 'shields-staging',
+        },
+        staticPreview: this.render({
+          state: 'success',
+        }),
+        documentation,
+      },
+    ]
+  }
+
+  static get defaultBadgeData() {
+    return { label: 'state' }
+  }
+
+  static render({ state }) {
+    const { label, message: build_message, color } = renderBuildStatusBadge({
+      label: 'state',
+      status: state,
+    })
+
+    let message = build_message
+
+    // renderBuildStatusBadge() does everything we need except it rewrites the
+    // message for status in 'greenStatuses' to 'passing'. This makes little
+    // sense in the context of a deployment. So rewrite it back to 'success'.
+    if (state == 'success') {
+      message = 'success'
+    }
+
+    return {
+      label,
+      message,
+      color,
+    }
+  }
+
+  async fetch({ user, repo, environment }) {
+    return await this._requestGraphql({
+      query: gql`
+        query($user: String!, $repo: String!, $environment: String!) {
+          repository(owner: $user, name: $repo) {
+            deployments(last: 1, environments: [$environment]) {
+              nodes {
+                latestStatus {
+                  state
+                }
+              }
+            }
+          }
+        }
+      `,
+      variables: { user, repo, environment },
+      schema,
+      transformErrors,
+    })
+  }
+
+  transform({ data }) {
+    if (data.repository.deployments.nodes.length == 0) {
```

Comment:
In what scenarios would this occur? Is it when the user specifies an invalid environment value, when there were no deployments to the specified environment, something else?

Reply:
That sounds a lot better :D One other corner case is when there is a deployment which did not receive a status update yet. I am not sure we can differentiate between "env not found" and "deployment without status". Also not sure it is worth the effort (given it is a rare corner case in normal operations).